### ISSUE TITLE
Incorrect output shape when using dilations and strides > 1 with depthwise conv

### ISSUE THREAD

**Issue by Hrayo712** [tensorflow/tensorflow#57128]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

tf 2.4 - v2.4.0-rc4-71-g582c8d236cb 2.4.0

### Custom Code

No

### OS Platform and Distribution

Linux Ubuntu 18.04.6 LTS

### Python version

3.8.0

### CUDA/cuDNN version

11.0

### Current Behaviour?

I am building a model which uses `strides` > 1 and `dilation_rate` > 1 in some DepthwiseConv2D layers.

However, with certain values, it seems like the output shape is incorrect.

Note: On the documentation, I see that using dilation_rate > 1 with strides > 1 is invalid, however, no runtime error is raised, in contrast with Conv2D, where the error is raised.

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

input = tf.keras.layers.Input(
            shape=(136, 136, 72),
            batch_size=1,
        )
x = tf.keras.layers.DepthwiseConv2D(
      kernel_size=3,
      strides=2,
      activation=None,
      use_bias=False,
      padding='valid',
      dilation_rate=(3, 3))(input)

print(f"output shape: {x.shape}")
```

The print statement shows: 
```
output shape: (1, 64, 64, 72)
```

Interestingly, when doing the following:
```
conv = tf.keras.layers.DepthwiseConv2D(
      kernel_size=3,
      strides=2,
      activation=None,
      use_bias=False,
      padding='valid',
      dilation_rate=(3, 3))

conv.compute_output_shape((1, 136, 136, 72))
```

The output reported is: 
```
(1, 65, 65, 72)
```

Interestingly, when using dilation_rate =(2, 2) instead, the outputs reported via these two examples, match.

Following the equation in: https://www.tensorflow.org/api_docs/python/tf/nn#notes_on_padding_2

for valid padding, the output shape is calculated as:

```
output = ceil((in_height - filter_height + 1) / stride_height)
```

where filter_height when using dilations is computed as:

```
filter_height: dilation*(filter_height - 1) + 1
```

Following these equations, the output shape as reported by `compute_output_shape` is correct. However, it doesnt match the shape reported when actually forwarding through the layer.

I also tried implementing this in Pytorch, and output shape reported matches the one reported by `compute_output_shape`

This might be related to this issue: https://github.com/keras-team/tf-keras/issues/79

Is this a bug ? I need to know how the output shape is computed, as I need to be able to calculate a certain amount of padding such to ensure the downsampling caused by the stride results in a certain shape.

Thanks in advance for your support!

### Relevant log output

_No response_</details>

**Comment by tilakrayal** [comment]
@Hrayo712,
TensorFlow v2.4 is not actively supported. Could you please update TensorFlow to the latest stable version v2.9.0 and check if you are facing the same issue. Thank you!

**Comment by Hrayo712** [comment]
@tilakrayal 

Just verified that the same behavior is present on TF v2.9.1

**Comment by tilakrayal** [comment]
@Hrayo712,
Thanks for opening this issue. Development of keras moved to another [repository](https://github.com/keras-team/keras/issues). 
 
 
Could you please post this issue on keras-team/keras [repo](https://github.com/keras-team/keras/issues).
To know more please refer:
https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999
Thank you!

**Comment by Hrayo712** [comment]
Thanks!

Reposted to: https://github.com/keras-team/keras/issues/16930

**Comment by hertschuh** [comment]
Reopening this instead of https://github.com/keras-team/keras/issues/16930 as the issue has to do with `tf.nn.depthwise_conv2d` and not the Keras wrapper on top of it.

This is also discussed in https://github.com/keras-team/keras/issues/16092

When strides and dilations are both > 1, convolutions produce results of different shapes depending on whether you're running on TPU or not.

Repro:
```python
depthwise_kernel = tf.zeros((3, 3, 72, 1))
input = tf.ones((1, 136, 136, 72))

with tf.device('/TPU:0'):
  res1 = tf.compat.v1.nn.depthwise_conv2d_native(
        input,
        depthwise_kernel,
        strides=(1, 2, 2, 1),
        padding='VALID',
        dilations=(1, 3, 3, 1),
        data_format='NHWC',
    )

print(res1.device)
print(res1.shape)

res2 = tf.compat.v1.nn.depthwise_conv2d(
        input,
        depthwise_kernel,
        strides=(1, 2, 2, 1),
        padding='VALID',
        rate=(3, 3),
        data_format='NHWC',
    )

print(res2.device)
print(res2.shape)
```
Output:
```
/job:worker/replica:0/task:0/device:TPU:0
(1, 65, 65, 72)
/job:worker/replica:0/task:0/device:CPU:0
(1, 64, 64, 72)
```
The formula for the output size based on the inputs, filter size, strides and dilation rate is the following (for each spatial dimension):
```
output_size = floor((input_size - (dilation_rate * (filter_size - 1) + 1) + stride) / stride)
```
This is indeed the formula that is implemented here [GetWindowedOutputSizeFromDimsV2](https://github.com/tensorflow/tensorflow/blob/b9862801b812d09c5b885680aa95d2a771729f43/tensorflow/core/framework/common_shape_fns.cc#L35)

And also here [GetWindowedOutputSizeVerboseV2](https://source.corp.google.com/piper///depot/google3/third_party/tensorflow/core/framework/kernel_shape_util.cc;l=20)

This is also how [Keras computes it](https://github.com/keras-team/keras/blob/master/keras/utils/conv_utils.py#L118)

However, the non-TPU implementation preprocesses the input and post-processes the output to take the dilation out and run a convolution without dilation, see the [with_space_to_batch call](https://github.com/tensorflow/tensorflow/blob/v2.11.0/tensorflow/python/ops/nn_impl.py#L887) and the [with_space_to_batch implementation](https://github.com/tensorflow/tensorflow/blob/v2.11.0/tensorflow/python/ops/nn_ops.py#L569)

```
intermediary_size = floor((input_size - 1) / dilation_rate) + 1
conv_output_size = floor((intermediary_size - filter_size + stride) / stride)
```
Note that there is also some extra padding added and removed in the process, using [this formula](https://github.com/tensorflow/tensorflow/blob/v2.11.0/tensorflow/python/ops/array_ops.py#L4064)

In the repro example:
```
filter_size = 3
stride = 2
dilation_rate = 3
input_size = 136
```
The "correct" formula produces `65`, which is what we see on TPU.

But with the intermediary, it produce `64`, which is what we see on CPU and GPU.

So applying the formula with the intermediary is not the same as the overall formula. The result is off by one half of the time.

I suspect the way to fix this would be to apply a different extra padding on the way in and the way out.

**Comment by tilakrayal** [comment]
@sachinprasadhs,
I was able to reproduce the issue on tensorflow [v2.9](https://colab.research.google.com/gist/tilakrayal/fc3bb093faa07f69c6e82f49cb1efdd9/untitled831.ipynb), [v2.11](https://colab.research.google.com/gist/tilakrayal/01952d120576ddb7108abad1625fbadb/untitled832.ipynb) and [nightly](https://colab.research.google.com/gist/tilakrayal/2075154ea5c2a250e9399e97ea0ba2f3/untitled833.ipynb). Kindly find the gist.

**Comment by tilakrayal** [comment]
@Hrayo712,
I tried to execute the mentioned code on tf-nightly and keras 3.0, the code was executed with the error which was intended. Whereas on the older tensorflow version, it was executed without providing the error. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/60e2650d015f2bedba224c4f4b748223/untitled1831.ipynb). Thank you!

**Comment by aniruthraj** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
